templates/tftmpl/integration_test.go

```diff
@@ -229,21 +229,26 @@ func TestRenderTFVarsTmpl(t *testing.T) {
 			defer cancel()
 
 			gld := testutils.CheckFile(t, true, tc.goldenFile, "")
-			retry := 0
+			var cachedContents string
 			for {
 				re, err := r.Run(tmpl, w)
 				require.NoError(t, err)
 
 				if re.Complete {
 					// there may be a race with the consul services registering
 					// let's retry once.
-					if (string(gld) != string(re.Contents)) && retry < 1 {
-						retry++
+					contents := string(re.Contents)
+					if (string(gld) != contents) && cachedContents == "" {
+						cachedContents = contents
 						continue
 					}
 					assert.Equal(t, string(gld), string(re.Contents))
 					break
 				}
+				if cachedContents != "" {
+					assert.Equal(t, string(gld), cachedContents)
+					t.Fail()
+				}
```

Comment:
It looks like this assert.Equal will always fail so long as `string(gld) != contents` on the first `re.Complete`. I might be missing something. I'm not sure that I'm following the intention of this if-block.

Also was wondering about using `t.Fail()`. I'm not familiar with it! From initial reading I see that it marks the test as failed. Is it doing some in addition to assert.Equal, which also marks failure?

Reply:
idk if the race condition still exists, so i kept the retry in there but perhaps since the word retry is not used it's not as clear. I updated it to be more clear.

There's a bug in the test where `re.Complete` is true but the rendered content does not match the expected. In this case, it'll attempt a retry and `r.Run()` again but the second time `re.Complete` returns false and then block indefinitely on `w.WaitCh` because there is no new data.

---

Assert marks that test as failed only when it returns. However assert does not stop the test from continuing, and in this case it'll continue and proceed on the wait + for loop.

Reply:
Ah! Gotcha. Thank you for all the details you shared around the `re.Complete` and `t.Fail`. I did a poor job of following the changes 🤦 . The 'retry' var you added in this iteration was helpful for me. Thanks!